File: web/lib/api-usage.ts
/**
 * api-usage — provider-agnostic infra-cost logger.
 *
 * Writes to the `api_usage_log` table. Distinct from
 * `lib/cost-tracking/tracker` (which writes per-user/per-campaign
 * rows to `generation_costs`). OpenAI campaign routes write to both:
 * `generation_costs` for billing attribution, `api_usage_log` for
 * unified observability across OpenAI + Searchapi + Apify.
 *
 * Server-only. Uses service-role client so RLS doesn't block inserts.
 *
 * Design:
 *   1. NEVER throw — cost logging is observability, not a transaction.
 *      Failures are console.warn'd and swallowed.
 *   2. fire-and-forget: callers can `void logApiCall(...)` or await.
 *   3. request_id is generated at log time as a UUID; idempotency
 *      within api_usage_log only (no cross-table use).
 */

import { randomUUID } from "node:crypto";
import { createClient, type SupabaseClient } from "@supabase/supabase-js";
import {
  calculateLLMCost,
  calculateImageCost,
  calculateTTSCost,
  calculateVideoCost,
} from "./cost-tracking/calculator";

export type ApiProvider = "openai" | "searchapi" | "apify";

export type ApiUnitType =
  | "tokens"
  | "searches"
  | "compute_units"
  | "characters"
  | "seconds"
  | "images";

export interface LogApiCallInput {
  provider: ApiProvider;
  operation: string;
  model_or_actor: string;
  units_consumed: number;
  unit_type: ApiUnitType;
  cost_usd: number;
  called_from: string;
  tenant_id?: string | null;
  request_id?: string;
  metadata?: Record<string, unknown>;
}

let cachedServiceClient: SupabaseClient | null = null;

function serviceClient(): SupabaseClient | null {
  if (cachedServiceClient) return cachedServiceClient;
  const url = process.env.NEXT_PUBLIC_SUPABASE_URL ?? process.env.SUPABASE_URL;
  const key =
    process.env.SUPABASE_SERVICE_ROLE_KEY ?? process.env.SUPABASE_SERVICE_KEY;
  if (!url || !key) return null;
  cachedServiceClient = createClient(url, key, {
    auth: { persistSession: false, autoRefreshToken: false },
  });
  return cachedServiceClient;
}

/**
 * Record a single API call. Always returns; never throws.
 *
 * Pass a `request_id` explicitly only when the upstream provider gives
 * you one (e.g. Apify run id) so retries dedupe. Otherwise we generate
 * a UUID here.
 */
export async function logApiCall(input: LogApiCallInput): Promise<void> {
  const client = serviceClient();
  if (!client) {
    console.warn("[api-usage] missing Supabase service env; skipping log");
    return;
  }

  const row = {
    provider: input.provider,
    operation: input.operation,
    model_or_actor: input.model_or_actor,
    units_consumed: input.units_consumed,
    unit_type: input.unit_type,
    cost_usd: input.cost_usd,
    request_id: input.request_id ?? randomUUID(),
    called_from: input.called_from,
    tenant_id: input.tenant_id ?? null,
    metadata: input.metadata ?? {},
  };

  try {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const db = client as any;
    const { error } = await db.from("api_usage_log").insert(row);
    if (error) {
      console.warn(`[api-usage] insert failed: ${error.message}`);
    }
  } catch (e) {
    const message = e instanceof Error ? e.message : String(e);
    console.warn(`[api-usage] insert threw: ${message}`);
  }
}

/* ──────────────────────────────────────────────────────────────────────── */
/* OpenAI helpers — convert usage shape to a logApiCall payload              */
/* ──────────────────────────────────────────────────────────────────────── */

export interface LogOpenAITokenCallInput {
  operation: string;
  model: string;
  input_tokens: number;
  output_tokens: number;
  called_from: string;
  tenant_id?: string | null;
  request_id?: string;
  metadata?: Record<string, unknown>;
}

export async function logOpenAITokenCall(
  input: LogOpenAITokenCallInput,
): Promise<void> {
  const cost_cents = calculateLLMCost("openai", input.model, {
    input_tokens: input.input_tokens,
    output_tokens: input.output_tokens,
  });
  await logApiCall({
    provider: "openai",
    operation: input.operation,
    model_or_actor: input.model,
    units_consumed: input.input_tokens + input.output_tokens,
    unit_type: "tokens",
    cost_usd: cost_cents / 100,
    called_from: input.called_from,
    tenant_id: input.tenant_id,
    request_id: input.request_id,
    metadata: {
      input_tokens: input.input_tokens,
      output_tokens: input.output_tokens,
      ...(input.metadata ?? {}),
    },
  });
}

export interface LogOpenAIImageCallInput {
  operation: string;
  model: string;
  image_count: number;
  called_from: string;
  tenant_id?: string | null;
  request_id?: string;
  metadata?: Record<string, unknown>;
}

export async function logOpenAIImageCall(
  input: LogOpenAIImageCallInput,
): Promise<void> {
  const cost_cents = calculateImageCost("openai", input.model, input.image_count);
  await logApiCall({
    provider: "openai",
    operation: input.operation,
    model_or_actor: input.model,
    units_consumed: input.image_count,
    unit_type: "images",
    cost_usd: cost_cents / 100,
    called_from: input.called_from,
    tenant_id: input.tenant_id,
    request_id: input.request_id,
    metadata: input.metadata,
  });
}

export interface LogOpenAITTSCallInput {
  operation: string;
  model: string;
  characters: number;
  called_from: string;
  tenant_id?: string | null;
  request_id?: string;
  metadata?: Record<string, unknown>;
}

export async function logOpenAITTSCall(
  input: LogOpenAITTSCallInput,
): Promise<void> {
  const cost_cents = calculateTTSCost("openai", input.model, input.characters);
  await logApiCall({
    provider: "openai",
    operation: input.operation,
    model_or_actor: input.model,
    units_consumed: input.characters,
    unit_type: "characters",
    cost_usd: cost_cents / 100,
    called_from: input.called_from,
    tenant_id: input.tenant_id,
    request_id: input.request_id,
    metadata: input.metadata,
  });
}

export interface LogOpenAIVideoCallInput {
  operation: string;
  model: string;
  seconds: number;
  called_from: string;
  tenant_id?: string | null;
  request_id?: string;
  metadata?: Record<string, unknown>;
}

export async function logOpenAIVideoCall(
  input: LogOpenAIVideoCallInput,
): Promise<void> {
  const cost_cents = calculateVideoCost("openai", input.model, input.seconds);
  await logApiCall({
    provider: "openai",
    operation: input.operation,
    model_or_actor: input.model,
    units_consumed: input.seconds,
    unit_type: "seconds",
    cost_usd: cost_cents / 100,
    called_from: input.called_from,
    tenant_id: input.tenant_id,
    request_id: input.request_id,
    metadata: input.metadata,
  });
}

/* ──────────────────────────────────────────────────────────────────────── */
/* Searchapi pricing — read flat rate from api_pricing_config                */
/* ──────────────────────────────────────────────────────────────────────── */

interface PricingRow {
  rate_per_unit_usd: number;
  monthly_quota_units: number | null;
}

let cachedSearchapiRate: { value: PricingRow; ts: number } | null = null;
const PRICING_CACHE_MS = 5 * 60 * 1000; // 5 minutes

/**
 * Fetch Searchapi rate + quota from `api_pricing_config`. Cached for
 * 5 minutes; fall back to Developer-plan defaults if the table is
 * unreachable (offline tests, transient DB error).
 */
export async function getSearchapiPricing(): Promise<PricingRow> {
  const FALLBACK: PricingRow = { rate_per_unit_usd: 0.0099, monthly_quota_units: 10000 };

  if (cachedSearchapiRate && Date.now() - cachedSearchapiRate.ts < PRICING_CACHE_MS) {
    return cachedSearchapiRate.value;
  }

  const client = serviceClient();
  if (!client) return FALLBACK;

  try {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const db = client as any;
    const { data, error } = await db
      .from("api_pricing_config")
      .select("rate_per_unit_usd, monthly_quota_units, effective_from")
      .eq("provider", "searchapi")
      .eq("unit_type", "searches")
      .lte("effective_from", new Date().toISOString().slice(0, 10))
      .order("effective_from", { ascending: false })
      .limit(1)
      .maybeSingle();

    if (error || !data) return FALLBACK;

    const value: PricingRow = {
      rate_per_unit_usd: Number(data.rate_per_unit_usd),
      monthly_quota_units: data.monthly_quota_units ?? null,
    };
    cachedSearchapiRate = { value, ts: Date.now() };
    return value;
  } catch {
    return FALLBACK;
  }
}

/* ──────────────────────────────────────────────────────────────────────── */
/* Searchapi wrapper                                                         */
/* ──────────────────────────────────────────────────────────────────────── */

export interface SearchapiGetInput {
  url: string;
  params: Record<string, string | number | undefined>;
  called_from: string;
  operation: string;
  tenant_id?: string | null;
}

/**
 * GET against Searchapi.io. Records one log row per successful response.
 * Throws on HTTP error (caller decides retry policy); does NOT log
 * failed calls (avoids paying for retries twice in the dashboard).
 */
export async function searchapiGet(input: SearchapiGetInput): Promise<unknown> {
  const url = new URL(input.url);
  for (const [k, v] of Object.entries(input.params)) {
    if (v !== undefined) url.searchParams.set(k, String(v));
  }

  const resp = await fetch(url.toString(), { method: "GET" });
  if (!resp.ok) {
    throw new Error(`Searchapi ${resp.status}: ${await resp.text()}`);
  }
  const json = await resp.json();

  const pricing = await getSearchapiPricing();
  void logApiCall({
    provider: "searchapi",
    operation: input.operation,
    model_or_actor: String(input.params.engine ?? "search"),
    units_consumed: 1,
    unit_type: "searches",
    cost_usd: pricing.rate_per_unit_usd,
    called_from: input.called_from,
    tenant_id: input.tenant_id,
    metadata: { engine: input.params.engine, q: input.params.q },
  });

  return json;
}

File: pipeline/lib/api_usage.py
"""
api_usage — provider-agnostic infra-cost logger for pipelines.

Mirrors web/lib/api-usage.ts. Writes to the `api_usage_log` Supabase
table. Distinct from `generation_costs`, which the web app populates
for per-user/per-campaign attribution.

Design:
  1. NEVER raises — cost logging is observability, not a transaction.
     Failures are printed and swallowed so they can't break a pipeline.
  2. Pass-through wrappers (`searchapi_get`, `run_apify_actor`) instrument
     calls automatically so individual pipelines don't have to remember
     to log.

Env:
  SUPABASE_URL / SUPABASE_SERVICE_KEY (or SUPABASE_SERVICE_ROLE_KEY):
    standard pipeline env. Reused via lib.pipeline helpers below.
  DRY_RUN: respected — skip inserts when set.
"""
from __future__ import annotations

import os
import time
import uuid
from typing import Any, Optional

import httpx

from .api_pricing import get_searchapi_pricing

SUPABASE_URL = os.environ.get("SUPABASE_URL", "")
SUPABASE_KEY = (
    os.environ.get("SUPABASE_SERVICE_KEY", "")
    or os.environ.get("SUPABASE_SERVICE_ROLE_KEY", "")
)
DRY_RUN = os.environ.get("DRY_RUN", "false").lower() in ("true", "1", "yes")

APIFY_API_BASE = "https://api.apify.com/v2"


def _headers() -> dict[str, str]:
    return {
        "apikey": SUPABASE_KEY,
        "Authorization": f"Bearer {SUPABASE_KEY}",
        "Content-Type": "application/json",
    }


def log_api_call(
    *,
    provider: str,
    operation: str,
    model_or_actor: str,
    units_consumed: float,
    unit_type: str,
    cost_usd: float,
    called_from: str,
    tenant_id: Optional[str] = None,
    request_id: Optional[str] = None,
    metadata: Optional[dict] = None,
) -> None:
    """Record a single API call. Never raises.

    Args:
      provider: 'openai' | 'searchapi' | 'apify'.
      unit_type: 'tokens' | 'searches' | 'compute_units' | 'characters' |
                 'seconds' | 'images'.
      called_from: pipeline module name (e.g. 'pipelines.ad_intel_daily').
      request_id: provider's run id when known (Apify run id), else UUID.
    """
    if DRY_RUN:
        return
    if not SUPABASE_URL or not SUPABASE_KEY:
        print("[api-usage] missing Supabase env; skipping log")
        return

    row = {
        "provider": provider,
        "operation": operation,
        "model_or_actor": model_or_actor,
        "units_consumed": units_consumed,
        "unit_type": unit_type,
        "cost_usd": cost_usd,
        "request_id": request_id or str(uuid.uuid4()),
        "called_from": called_from,
        "tenant_id": tenant_id,
        "metadata": metadata or {},
    }

    try:
        resp = httpx.post(
            f"{SUPABASE_URL}/rest/v1/api_usage_log",
            headers=_headers(),
            json=row,
            timeout=15,
        )
        if resp.status_code >= 400:
            print(f"[api-usage] insert failed {resp.status_code}: {resp.text[:200]}")
    except Exception as e:  # noqa: BLE001 — observability path must not crash pipelines
        print(f"[api-usage] insert threw: {e}")


# ---------------------------------------------------------------------------
# Searchapi wrapper — one log row per successful GET
# ---------------------------------------------------------------------------


def searchapi_get(
    url: str,
    params: dict[str, Any],
    *,
    called_from: str,
    operation: str,
    timeout: int = 30,
) -> httpx.Response:
    """GET against Searchapi.io and log a usage row on 2xx.

    Returns the raw httpx.Response so callers can keep their existing
    .json() / .raise_for_status() flow. Failures are not logged
    (don't pay for retries twice in the dashboard).
    """
    resp = httpx.get(url, params=params, timeout=timeout)
    if 200 <= resp.status_code < 300:
        rate = get_searchapi_pricing()
        log_api_call(
            provider="searchapi",
            operation=operation,
            model_or_actor=str(params.get("engine", "search")),
            units_consumed=1,
            unit_type="searches",
            cost_usd=rate["rate_per_unit_usd"],
            called_from=called_from,
            metadata={"engine": params.get("engine"), "q": params.get("q")},
        )
    return resp


# ---------------------------------------------------------------------------
# Apify wrapper — cost is read from the run's terminal-state usage field
# ---------------------------------------------------------------------------


def run_apify_actor(
    actor_id: str,
    actor_input: dict,
    *,
    label: str,
    called_from: str,
    apify_token: str,
    run_timeout_seconds: int = 300,
    poll_interval: int = 5,
) -> list[dict]:
    """Run an Apify actor end-to-end and log the cost reported by Apify.

    Cost is captured from `usage.totalUsageUsd` on the run record
    once it reaches a terminal state — same poll loop we already do,
    so no extra latency.

    Returns the dataset items the actor produced (or [] if none).
    Raises on failure / timeout, exactly like the legacy helper.
    """
    run_resp = httpx.post(
        f"{APIFY_API_BASE}/acts/{actor_id}/runs",
        params={"token": apify_token},
        json=actor_input,
        timeout=60,
    )
    run_resp.raise_for_status()
    run_data = run_resp.json().get("data", {})
    run_id = run_data.get("id")
    if not run_id:
        raise ValueError(f"Apify actor run id missing for {label}")

    started = time.time()
    while True:
        status_resp = httpx.get(
            f"{APIFY_API_BASE}/actor-runs/{run_id}",
            params={"token": apify_token},
            timeout=60,
        )
        status_resp.raise_for_status()
        run_info = status_resp.json().get("data", {})
        status = run_info.get("status")

        if status == "SUCCEEDED":
            _log_apify_run(run_info, actor_id, called_from, label)
            dataset_id = run_info.get("defaultDatasetId")
            if not dataset_id:
                return []
            items_resp = httpx.get(
                f"{APIFY_API_BASE}/datasets/{dataset_id}/items",
                params={"token": apify_token},
                timeout=120,
            )
            items_resp.raise_for_status()
            return items_resp.json()

        if status in {"FAILED", "ABORTED", "TIMED-OUT"}:
            # Still log so the dashboard surfaces wasted spend on failed runs.
            _log_apify_run(run_info, actor_id, called_from, label)
            raise RuntimeError(f"Apify actor {label} ended with status={status}")

        if time.time() - started >= run_timeout_seconds:
            raise TimeoutError(
                f"Apify actor {label} exceeded {run_timeout_seconds}s timeout"
            )

        time.sleep(poll_interval)


def _log_apify_run(
    run_info: dict,
    actor_id: str,
    called_from: str,
    label: str,
) -> None:
    """Read cost off an Apify run record and log it."""
    usage = run_info.get("usage") or {}
    usage_totals = run_info.get("usageTotalUsd")
    cost_usd: float
    if isinstance(usage_totals, (int, float)):
        cost_usd = float(usage_totals)
    elif isinstance(usage.get("totalUsageUsd"), (int, float)):
        cost_usd = float(usage["totalUsageUsd"])
    else:
        cost_usd = 0.0

    compute_units = usage.get("ACTOR_COMPUTE_UNITS") or run_info.get("computeUnits") or 0

    log_api_call(
        provider="apify",
        operation=label,
        model_or_actor=actor_id,
        units_consumed=float(compute_units),
        unit_type="compute_units",
        cost_usd=cost_usd,
        called_from=called_from,
        request_id=str(run_info.get("id") or uuid.uuid4()),
        metadata={
            "status": run_info.get("status"),
            "dataset_id": run_info.get("defaultDatasetId"),
            "started_at": run_info.get("startedAt"),
            "finished_at": run_info.get("finishedAt"),
        },
    )
